### PR TITLE
ASoC: SOF: topology: add support to smart amplifier

### PR DESCRIPTION
**Pull request by keyonjie**
Add smart amplifier component support, which is designed as another new
type of process component and used for Dynamic Speaker Management.

Signed-off-by: Keyon Jie <yang.jie@linux.intel.com>

### CODE REVIEW

**Comment by keyonjie** [comment]
The corresponding FW part PR is here:
https://github.com/thesofproject/sof/pull/1896

**Comment by plbossart** [comment]
I think this is a dangerous path. the 'smart-amp' is by no means generic and the code is not available, it's 3rd-party IP and it should be classified as such. This is not a full-blown process maintained in the SOF tree. At the very least it should be an ID that belongs in the 3rd party range

@lgirdwood what's your take on this?

**Comment by keyonjie** [comment]
@plbossart Thanks for reviewing.

I am not sure if your "dangerous" means it could break the amplifier with bad algorithm, or adding more and more new component ID is not the correct path, but let me try to add some information about what we have been doing. Per my understanding, the FW team has been using the model for 3rd-party IP development like this:
1. create a dummy(or simple) algorithm plus the component shell, integrate them to the whole SOF stack.
2. provide standard APIs for integrating 3rd-party algorithms, e.g.
 a. the detect_func() in detector component for KWD, where samples and size are provided and detected or not result is expected from the return
 b. the smart_amp_process() in smart_amp component, where the playback input and the feedback input samples are provided, the output samples are expected to be written to the sink buffer.
3. Our release will focus on the "generic" framework for the feature only, we are implement these dummy sample algorithm, creating topology files, trying to validate and make sure the sequence run as in the realword scenarios.

@bkokoszx @mrajwa please correct me if I have written something wrong.

**Comment by ranj063** [comment]
> I think this is a dangerous path. the 'smart-amp' is by no means generic and the code is not available, it's 3rd-party IP and it should be classified as such. This is not a full-blown process maintained in the SOF tree. At the very least it should be an ID that belongs in the 3rd party range
> 
> @lgirdwood what's your take on this?

@plbossart I agree that this isnt a scalable solution for supporting 3rd party modules. Do we have a process for adding new modules?

**Comment by plbossart** [comment]
> 1. create a dummy(or simple) algorithm plus the component shell, integrate them to the whole SOF stack.

That makes it impossible to know if the component is implemented with the dummy test version or the real one... Something is missing here really.

**Comment by keyonjie** [comment]
> > 1. create a dummy(or simple) algorithm plus the component shell, integrate them to the whole SOF stack.
> 
> That makes it impossible to know if the component is implemented with the dummy test version or the real one... Something is missing here really.

By default we are using dummy test version in upstream. We plan to add config item for 3rd-party version selection.

**Comment by kv2019i** [comment]
@keyonjie How does this relate to:
https://github.com/orgs/thesofproject/teams/sof-developers/discussions/32 
... you started this already three weeks back. But here in #1971 we are still adding a component type in the old way. Is this an intermediate thing, or is the proposal to add smart amp as a "well known" processing type?

**Comment by keyonjie** [comment]
> @keyonjie How does this relate to:
> https://github.com/orgs/thesofproject/teams/sof-developers/discussions/32
> ... you started this already three weeks back. But here in #1971 we are still adding a component type in the old way. Is this an intermediate thing, or is the proposal to add smart amp as a "well known" processing type?

Yes they have relationship, the discussion/32 is something like framework refinement, but I had no time to do that yet, it need effort on both FW and driver side together.

**Comment by plbossart** [comment]
I am personally not comfortable with merging this
a) internal releases already took it without reviews, so there's no urgency in topic/sof-dev
b) we don't know what the next steps are
c) it's again a piecemeal approach with incremental patches and no design document.

Call me a grumpy old person if you want, but it's a recurring pattern really. There was plenty of time to do the right thing and we didn't.

**Comment by keyonjie** [comment]
> I am personally not comfortable with merging this
> a) internal releases already took it without reviews, so there's no urgency in topic/sof-dev
> b) we don't know what the next steps are
> c) it's again a piecemeal approach with incremental patches and no design document.
> 
> Call me a grumpy old person if you want, but it's a recurring pattern really. There was plenty of time to do the right thing and we didn't.

Hi @plbossart  There is discussion and extra work to implement the UUID in the driver side, we need to align on that before we can refine the generic processing components.

**Comment by lgirdwood** [comment]
@plbossart @keyonjie @kv2019i set to MINOR 15 like FW

**Comment by lgirdwood** [comment]
> @plbossart @keyonjie @kv2019i set to MINOR 15 like FW

NAK 15, now at 16. @keyonjie please use MINOR 16.

**Comment by keyonjie** [comment]
> > @plbossart @keyonjie @kv2019i set to MINOR 15 like FW
> 
> NAK 15, now at 16. @keyonjie please use MINOR 16.

Thanks for the information @lgirdwood .

So should we add explicit ABI bump for this change or the MINOR 16 is for update with several other changes? I am not seeing the 16 bumping in the FW corresponding PR https://github.com/thesofproject/sof/pull/1896.

**Comment by keyonjie** [comment]
@lgirdwood @plbossart @kv2019i just checked the ABI MINOR in dirver side is already 16.

**Comment by kv2019i** [comment]
Update on this PR: the proposal is to use same approach as with EQ, KPB and others for Smart Amp.
The more generic solution to describe and control "effect" type components, is not ready yet (work on this continues in https://github.com/thesofproject/sof/issues/769 and https://github.com/orgs/thesofproject/teams/sof-developers/discussions/32).

File: sound/soc/intel/boards/sof_maxim_common.c [hunk 1]
```diff
@@ -47,11 +47,11 @@ static int max98373_hw_params(struct snd_pcm_substream *substream,
 	for_each_rtd_codec_dais(rtd, j, codec_dai) {
 		if (!strcmp(codec_dai->component->name, MAX_98373_DEV0_NAME)) {
 			/* DEV0 tdm slot configuration */
-			snd_soc_dai_set_tdm_slot(codec_dai, 0x30, 3, 8, 16);
+			snd_soc_dai_set_tdm_slot(codec_dai, 0x3, 3, 8, 32);
```
Comment (RanderWang) [inline sound/soc/intel/boards/sof_maxim_common.c]: unify code style to 0x3, 0x3, 8, 32 ?  tx_mask and rx_mask should be in the same style.

Comment (keyonjie) [reply sound/soc/intel/boards/sof_maxim_common.c]: sorry, I just pushed wrong local change recently, this is for debug only so should not go to here :)

Comment (keyonjie) [reply sound/soc/intel/boards/sof_maxim_common.c]: removed. @RanderWang

**Comment by kv2019i** [comment]
@plbossart wrote:
> let's hope this is the last one we add this way - but I wouldn't hold my breath.

And you were right of course. Although this was so quick that you could have even held your breath for real while waiting for this one ;D
https://github.com/thesofproject/linux/pull/2030

More seriously, I think it is now pretty evident that we need to invest in the infra in this area. I.e. time to proceed with https://github.com/orgs/thesofproject/teams/sof-developers/discussions/32

**Comment by keyonjie** [comment]
Thanks @plbossart @kv2019i , yes, that's true, I will put effort to the UUID implementation on driver side after freed from the release.

**Comment by plbossart** [comment]
@keyonjie and @sebcarlucci can you please chat and figure out which one of these two components goes in first and what ABI level this is.

@lgirdwood FYI this is a mess

**Comment by keyonjie** [comment]
> @keyonjie and @sebcarlucci can you please chat and figure out which one of these two components goes in first and what ABI level this is.

@sebcarlucci is it OK for you to let this smart amplifier one go first, it has been hold for quite long time and it is actually waited from the product.

For the ABI version, I think @lgirdwood and @kv2019i has aligned to use Minor 16 for it? @kv2019i has initiated an discussion about ABI refinement in https://github.com/orgs/thesofproject/teams/sof-developers/discussions/35, but I am not sure will that impact this PR actually.

> 
> @lgirdwood FYI this is a mess

**Comment by sebcarlucci** [comment]
> @sebcarlucci is it OK for you to let this smart amplifier one go first, it has been hold for quite long time and it is actually waited from the product.

@keyonjie Sounds good to me

**Comment by keyonjie** [comment]
> > @sebcarlucci is it OK for you to let this smart amplifier one go first, it has been hold for quite long time and it is actually waited from the product.
> 
> @keyonjie Sounds good to me

Thanks.

**Comment by kv2019i** [comment]
There were errors on one machine in CI: https://sof-ci.01.org/linuxpr/PR1971/build3638/devicetest/
... but given the commit contents, I can't see how it could be related. 

So basicly, I could merge this, but I'm a bit puzzled I see no approvals for the FW change https://github.com/thesofproject/sof/pull/1896 .. what's up with that?

**Comment by kv2019i** [comment]
The FW change is still missing reviews, not merging this before it's at least approved.

**Comment by keyonjie** [comment]
@kv2019i the FW one is approved now: https://github.com/thesofproject/sof/pull/1896

Let's merge it?

**Comment by keyonjie** [comment]
@kv2019i the FW one is approved now: https://github.com/thesofproject/sof/pull/1896

Let's merge it?

**Comment by keyonjie** [comment]
@kv2019i the FW one is approved now: https://github.com/thesofproject/sof/pull/1896

Let's merge it?

**Comment by kv2019i** [comment]
FW patch approved, merging this as well.

**Pull request merged**